### PR TITLE
Add % current file register

### PR DESCRIPTION
**Pull request by pickfire**
Useful to open files relative to current file.

[asciicast](https://asciinema.org/a/hOpVj7GcalGP6SvfRI9H9UHRn)

The experience is worse especially on files that have huge paths, like python virtual environment, this allows browsing relative files easier, while being a lot easier to implement compared to find file. https://github.com/helix-editor/helix/pull/2412

To create new file from current file directory.

    :o <C-r>%<C-w>new-file<ret>

Compared to previous method (mouse copy-paste or manually type - and I usually manually type 1-7 directories)

    :o <mouse select file name><middle click paste><manually type full path if truncated><C-w>new-file<ret>

### CODE REVIEW

**Comment by archseer** [comment]
Duplicate of https://github.com/helix-editor/helix/pull/3008 was replaced by https://github.com/helix-editor/helix/pull/3393

**Comment by pickfire** [comment]
I think #3393 is slightly different and a lot harder to use compared to this scenario, the user has to type `%val{dirname}` (16 keystrokes) as compared to `<C-r>%<C-w>` (6 keystrokes).

The one implemented here is slightly similar to what @QiBaobin implemented, except being pull based (when user select registers) rather than push based (when user change view).

**Comment by the-mikedavis** [comment]
The register could be useful for prompts and pickers where variable substitution might not be done (it might not make sense to replace variables if the prompt is a regex for example).

Kakoune has a special register for this as well: https://github.com/mawww/kakoune/blob/master/doc/pages/registers.asciidoc#special-registers

File: helix-view/src/editor.rs [hunk 1]
```diff
@@ -1469,6 +1469,17 @@ impl Editor {
...
     }
+
+    /// Update read-only registers.
+    pub fn update_registers(&mut self) {
+        let doc = doc!(self);
+        if let Some(path) = doc.relative_path() {
+            self.registers
+                .write('%', vec![path.to_string_lossy().to_string()]);
+        } else {
+            self.registers.clear('%');
```
Comment (the-mikedavis) [inline helix-view/src/editor.rs:1480]: To match kakoune we could write the scratch buffer name into the register when it doesn't have a path

**Review by the-mikedavis — COMMENTED** [review]
Kakoune has an interesting way of passing state like the current buffer name to register functions: registers are defined with an anonymous function that takes `Context` and returns the value of the register. This is helpful for other special registers like `0..9` acting as regex captures. We don't need to emulate that exactly but I'm curious if we can implement this in a way that's easier to add the other special registers later

File: helix-view/src/editor.rs (hunk 1 above)
Comment (pickfire) [reply helix-view/src/editor.rs:1480]: Yeah, I guess that's fine too.

File: helix-view/src/register.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+/// Currently just wraps a `BTreeMap` of `Register`s
+#[derive(Debug)]
+pub struct Registers {
+    inner: BTreeMap<char, Register>,
+}
+
+impl Registers {
+    pub fn read(&self, doc: &Document, name: char) -> Option<Vec<String>> {
```
Comment (pickfire) [inline helix-view/src/register.rs:27]: @the-mikedavis I did what you mentioned but now I don't like the API to always return an owned value, we need to keep cloning.

**Pull request closed without merging**